File: 序号/1123. 最深叶节点的最近公共祖先.py
class Solution:
    def lcaDeepestLeaves(self, root: Optional[TreeNode]) -> Optional[TreeNode]:
        def dfs(node:TreeNode):
            if node is None:return 0,None
            leftDepth,lnode=dfs(node.left)
            rightDepth,rnode=dfs(node.right)
            if leftDepth==rightDepth:
                return leftDepth+1,node
            elif leftDepth>rightDepth:
                return leftDepth+1,lnode
            else:
                return rightDepth+1,rnode
        return dfs(root)[1]

File: 序号/1123. 最深叶节点的最近公共祖先.kt
/**
 * Example:
 * var ti = TreeNode(5)
 * var v = ti.`val`
 * Definition for a binary tree node.
 * class TreeNode(var `val`: Int) {
 *     var left: TreeNode? = null
 *     var right: TreeNode? = null
 * }
 */
class Solution {

    private fun dfs(node:TreeNode?):Int{
        if(node==null) return 0
        return max(dfs(node?.left),dfs(node?.right))+1
    }
    fun lcaDeepestLeaves(root: TreeNode?): TreeNode? {
        // 左右深度相等，当前就是共同祖先，否则谁更深就递归更深的
        if(root==null) return null
        val l = dfs(root?.left)
        val r = dfs(root?.right)
        if(l==r) return root
        if(l>r) return lcaDeepestLeaves(root?.left)
        return lcaDeepestLeaves(root?.right)
    }
}